model/host/host.go

```diff
@@ -3370,3 +3475,233 @@ func (h *Host) UnsetPersistentDNSInfo(ctx context.Context) error {
+// indefinitely).
+var SleepScheduleSentinelTime = time.Date(10000, 0, 0, 0, 0, 0, 0, time.UTC)
+
+// GetNextScheduledStopTime returns the next time a host should be
+// stopped according to its sleep schedule.
+func (h *Host) GetNextScheduledStopTime(now time.Time) (time.Time, error) {
+	if h.SleepSchedule.PermanentlyExempt || h.SleepSchedule.ShouldKeepOff {
+		return SleepScheduleSentinelTime, nil
+	}
+
+	if err := h.SleepSchedule.Validate(); err != nil {
+		return time.Time{}, errors.Wrap(err, "invalid sleep schedule")
+	}
+
+	userTimeZone, err := time.LoadLocation(h.SleepSchedule.TimeZone)
+	if err != nil {
+		return time.Time{}, errors.Wrapf(err, "loading user time zone '%s'", h.SleepSchedule.TimeZone)
+	}
+
+	var calculateTimeAfter time.Time
+	if !utility.IsZeroTime(h.SleepSchedule.TemporarilyExemptUntil) {
+		calculateTimeAfter = h.SleepSchedule.TemporarilyExemptUntil
+	} else if !utility.IsZeroTime(h.SleepSchedule.NextStartTime) {
+		// If the next start time is known, do not try to stop the host again
+		// until after it's scheduled to start back up. This is a necessary
+		// safeguard against a rare edge case where the host is stopped for the
+		// sleep schedule but the user manually starts the host back up again in
+		// between a daily stop time and a whole weekday off.
+		//
+		// As an illustrative example, say the user sets a schedule to stop
+		// every day from 10 pm to 6 am, and also has Saturday and Sunday off.
+		// It's Friday at 11:30 PM and their host is currently stopped for the
+		// sleep schedule. They then manually start up their host. If the next
+		// stop time was calculated to be Saturday at midnight (the next
+		// scheduled whole day off), they may see their host abruptly shut down
+		// at midnight in the middle of working on it. A more graceful behavior
+		// would be to avoid stopping the host again until the next time the
+		// user would have expected their host to wake up (i.e. Monday at 6 am).
+		//
+		// To avoid this unfortunate scenario, once the host has been stopped
+		// for its sleep schedule, it should not try to stop the host again
+		// until after the next time the schedule would typically start it back
+		// up.
+		calculateTimeAfter = h.SleepSchedule.NextStartTime
+	} else {
+		calculateTimeAfter = now
+	}
+	// It's critical to do all the time calculations in the user's time zone and
+	// not in UTC. If the time calculations were all in UTC, they wouldn't
+	// account for timezone-specific quirks like daylight savings time changes.
+	calculateTimeAfter = calculateTimeAfter.In(userTimeZone)
+
+	// Partition the weekdays into the first day in a series of whole days off
+	// (if any) and the weekdays when the daily sleep schedule applies (if any).
+	var dailyWeekdays []string
+	var firstContiguousWeekdayOff []string
+	for _, weekday := range []time.Weekday{time.Sunday, time.Monday, time.Tuesday, time.Wednesday, time.Thursday, time.Friday, time.Saturday} {
```

Comment:
We can make this a constant since it's used twice

Reply:
Sure, I made some constants in utility and used them here.